### PR TITLE
feat: add DOC converter to MarkItDown

### PR DESCRIPTION
**Pull request by l-lumin**
Issue #23 

This PR adds a `DocConverter` class to convert DOC files into Markdown. It integrates the converter into the MarkItDown class and includes tests from #37  to verify the functionality

The conversion need use LibreOffice in headless mode.

### CODE REVIEW

**Comment by afourney** [comment]
Thanks for the PR. I agree we want legacy doc support. 

Let me think about this one a little. I'm not sure calling headless libreoffice (via shell) is the ideal approach here, and there may be other constraints. I know we do it for exiftool, but I'm not thrilled with that...

One nice thing is that the tools supports registering converters after initialization, similar to: 
https://github.com/microsoft/markitdown/blob/925c4499f72757abcf6cb521ee10e4844967af3d/src/markitdown/_markitdown.py#L1269C1-L1287C1

So in theory you could do:

```python
class DocConverter(HtmlConverter):
    ...
    def convert(self, local_path, **kwargs) -> Union[None, DocumentConverterResult]:
        ...

markitdown = MarkItDown()
markitdown.register_page_converter(DocConverter())
```

We could even imagine a little ecosystem of converters popping up on pypi, which would allow unrestricted growth in formats.

@gagb what do you think?

File: src/markitdown/_markitdown.py [hunk 1]
```diff
@@ -712,6 +713,46 @@ def convert(self, local_path, **kwargs) -> Union[None, DocumentConverterResult]:
...
+        try:
+            result = subprocess.run(libreoffice_command, capture_output=True, text=True)
+            output_file = outdir / (local_path.stem + ".txt")
+            with open(output_file, "r", encoding="utf-8") as f:
+                output_content = f.read()
+            output_file.unlink(missing_ok=True)
+            result = self._convert(output_content)
+            return result
+
+        except Exception as _:
```
Comment (gagb) [inline src/markitdown/_markitdown.py:752]: This part catches exceptions silently, can we update this? Maybe print warning?

File: src/markitdown/__init__.py [hunk 2]
```diff
@@ -2,9 +2,15 @@
...
-from ._markitdown import MarkItDown, FileConversionException, UnsupportedFormatException
+from ._markitdown import (
+    MarkItDown,
+    FileConversionException,
+    UnsupportedFormatException,
+    DocConverter,
+)
 
 __all__ = [
+    "DocConverter",
```
[the remark this answers is not in the document]
Comment (l-lumin) [reply src/markitdown/__init__.py:13]: Edited based on the comment from afourney, I can revert it

**Comment by gagb** [comment]
@aviral-bhardwaj can you please review this PR as well since you made a similar PR?

**Comment by gagb** [comment]
> Thanks for the PR. I agree we want legacy doc support.
> 
> Let me think about this one a little. I'm not sure calling headless libreoffice (via shell) is the ideal approach here, and there may be other constraints. I know we do it for exiftool, but I'm not thrilled with that...
> 
> One nice thing is that the tools supports registering converters after initialization, similar to: https://github.com/microsoft/markitdown/blob/925c4499f72757abcf6cb521ee10e4844967af3d/src/markitdown/_markitdown.py#L1269C1-L1287C1
> 
> So in theory you could do:
> 
> ```python
> class DocConverter(HtmlConverter):
>     ...
>     def convert(self, local_path, **kwargs) -> Union[None, DocumentConverterResult]:
>         ...
> 
> markitdown = MarkItDown()
> markitdown.register_page_converter(DocConverter())
> ```
> 
> We could even imagine a little ecosystem of converters popping up on pypi, which would allow unrestricted growth in formats.
> 
> @gagb what do you think?

I agree with your concern, but the PR
- adds functionality that is useful
- does not regress previous features

Could the refactor that you suggest be done in a different PR?

**Comment by aviral-bhardwaj** [comment]
> @aviral-bhardwaj can you please review this PR as well since you made a similar PR?

Will check later ,it is midnight in india 🇮🇳

**Review by StanFromIreland — APPROVED** [review]

File: src/markitdown/_markitdown.py [hunk 3]
```diff
@@ -712,6 +713,49 @@ def convert(self, local_path, **kwargs) -> Union[None, DocumentConverterResult]:
...
+            return None
+
+        if not (soffice := shutil.which("soffice")):
+            return None
+
+        local_path = Path(local_path)
+        outdir = local_path.parent
+        result = None
+
+        libreoffice_command = [
```
Comment (gagb) [inline src/markitdown/_markitdown.py:734]: Should we be telling users to install these libraries?

Comment (gagb) [reply src/markitdown/_markitdown.py:734]: Or install them for them?

Comment (l-lumin) [reply src/markitdown/_markitdown.py:737]:
```py
        if not (soffice := shutil.which("soffice")):
            return DocumentConverterResult(
                title=None,
                text_content="[ERROR] LibreOffice not found. Please install LibreOffice and try again.",
            )
```

added above

**Comment by aviral-bhardwaj** [comment]
Hi @l-lumin and @gagb,

This type of error should not be displayed to the end user. As the end user, they should not be responsible for installing additional libraries like LibreOffice on top of MarkitDown.

Is there a way to include LibreOffice as part of the installation process for our MarkitDown library? This way, when a user installs our library, everything will be set up, and they can start using it seamlessly without any extra steps.

**Comment by l-lumin** [comment]
> Hi @l-lumin and @gagb,
> 
> 
> 
> This type of error should not be displayed to the end user. As the end user, they should not be responsible for installing additional libraries like LibreOffice on top of MarkitDown.
> 
> 
> 
> Is there a way to include LibreOffice as part of the installation process for our MarkitDown library? This way, when a user installs our library, everything will be set up, and they can start using it seamlessly without any extra steps.

Include LibreOffice as part of installation process isn't ideal because can't account for all systems, which could cause issues.

I think do this will better
- Error message to guide user 
- Add LibreOffice to Dockerfile (I will create PR for Docker once this is merged)

**Comment by aviral-bhardwaj** [comment]
all are fine but test cases are failing in my system not sure why , attaching error file here @l-lumin , can you please have a look at this 

[I-lumin-test-code error.pdf](https://github.com/user-attachments/files/18207685/I-lumin-test-code.error.pdf)

**Comment by l-lumin** [comment]
> all are fine but test cases are failing in my system not sure why , attaching error file here @l-lumin , can you please have a look at this
> 
> [I-lumin-test-code error.pdf](https://github.com/user-attachments/files/18207685/I-lumin-test-code.error.pdf)

what environment are you running on? I checked your repository, and it is 16 commits ahead of microsoft/markitdown. It also includes failed tests

my test log

```sh
root@450635eb267c:/workspaces/markitdown# hatch test
================================================================================= test session starts ==================================================================================
platform linux -- Python 3.13.1, pytest-8.3.4, pluggy-1.5.0
rootdir: /workspaces/markitdown
configfile: pyproject.toml
plugins: rerunfailures-14.0, mock-3.14.0, anyio-4.7.0, xdist-3.6.1
collected 7 items                                                                                                                                                                      

tests/test_markitdown.py ..s...s                                                                                                                                                 [100%]

=================================================================================== warnings summary ===================================================================================
../../root/.local/share/hatch/env/virtual/markitdown/2jxz10qj/hatch-test.py3.13/lib/python3.13/site-packages/speech_recognition/__init__.py:7
  /root/.local/share/hatch/env/virtual/markitdown/2jxz10qj/hatch-test.py3.13/lib/python3.13/site-packages/speech_recognition/__init__.py:7: DeprecationWarning: aifc was removed in Python 3.13. Please be aware that you are currently NOT using standard 'aifc', but instead a separately installed 'standard-aifc'.
    import aifc

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================================================================= 5 passed, 2 skipped, 1 warning in 18.22s =======================================================================
root@450635eb267c:/workspaces/markitdown# 
```

**Comment by aviral-bhardwaj** [comment]
yes yes mine is window system thats why it is showing error

**Comment by aviral-bhardwaj** [comment]
@gagb please have a final look , this is fine from my end

File: src/markitdown/_markitdown.py [hunk 4]
```diff
@@ -712,6 +713,52 @@ def convert(self, local_path, **kwargs) -> Union[None, DocumentConverterResult]:
...
+    def convert(self, local_path, **kwargs) -> Union[None, DocumentConverterResult]:
+        # Bail if not a DOC
+        extension = kwargs.get("file_extension", "")
+        if extension.lower() != ".doc":
+            return None
+
+        if not (soffice := shutil.which("soffice")):
+            return DocumentConverterResult(
+                title=None,
+                text_content="[ERROR] LibreOffice not found. Please install LibreOffice and try again.",
```
Comment (gagb) [inline src/markitdown/_markitdown.py:730]:
```suggestion
                text_content="[ERROR] This feature requires the 'soffice' command from LibreOffice. Please install LibreOffice. https://www.libreoffice.org/get-help/install-howto/.",
```

**Comment by afourney** [comment]
Again, I am supportive of this effort, and appreciate the contribution, but I'm not sold on calling out to LibreOffice via shell to accomplish this conversion -- especially in the default installation. It works, but it's a pretty heavyweight binary dependency, and brings other baggage. 

Also, if LibreOffice is available, why not use it for PPT and other formats as well.

I think, very soon, I would like to create some sort of plugin ecosystem where you could do:

```bash
pip install markitdown
pip install markitdown-soffice
```

And then get all the libreOffice converters (perhaps detected and incorporated automatically). This would allow you, or anyone, to add alternatives or new file formats, and host and update them at your own convenience. 

However, I'm still working out the details for this.

I would like to close this PR for now, but come back to it as soon as we have the details worked out. Again, I am appreciative of this effort -- this is not a reflection on the code or anything, but rather a design decision on our end that has yet to be worked out.

**Comment by dridk** [comment]
I just start a discussion here about the support of the doc format. 
https://github.com/microsoft/markitdown/pull/153 

I'd rather encourage people to use apache tika for doc conversion than openoffice. 
But the best thing would be for microsoft to provide a native doc parser.  Because it would be really funny if a library made by microsoft had to use openoffice to read microsoft doc files.

**Pull request closed without merging**